### DISCUSSION TITLE
Unspecified cyclical build dependencies

### DISCUSSION THREAD

**Discussion by jbscherrer** [bonsai-rx/bonsai#1265] [Q&A]
I'm trying to write a code block that repeatedly takes in a series of images from a video stream, performs some operations on the image series, compares the result to a buffer containing the past N results, and then outputs a value that is used to update the buffer. I'm trying to use a BehaviorSubject to perform this operation, however I'm getting this error: `The workflow contains unspecified build dependencies`. I've attached a toy workflow that illustrates the problem.

[cyclical_dependencies.zip](https://github.com/bonsai-rx/bonsai/files/10828995/cyclical_dependencies.zip)

It seems the issue is that I have a SubscribeSubject in SelectMany that is subscribed to a BehaviorSubject that is periodically updated by the output of that same SelectMany. Why isn't it possible to do this? My understanding is that every instantiation of the SelectMany will get whatever value was last published to the BehaviorSubject from SubscribeSubject, and will then update its value after terminating.

**Comment by RoboDoig** [comment]
I think the issue here is that the definition of the BehaviorSubject depends on its own subscription. When constructing workflows in Bonsai there is some analysis happening that checks the workflow is valid. Bonsai needs to know what kind of data the BehaviorSubject will output, to determine this it looks through the inputs that lead to the BehaviorSubject. One of those inputs from the SelectMany is itself defined by a subscription to the BehaviorSubject so we end up in a loop where we can't know what the type of the SelectMany is without the BehaviorSubject and we can't know the type of the BehaviorSubject without knowing the type of the SelectMany.

I'm not sure I've understood your problem 100% but maybe you could do something like this?

![image](https://user-images.githubusercontent.com/11314305/221319514-e8ac8fcd-81f1-44b8-8ed6-efe213c8969d.png)

The first 4 nodes define what the data type of the 'previous value' will be. You then create a PropertySource from this which is a BehaviourSubject called 'PreviousValue'. Create a Multicast from the first line of nodes to initialize this BehaviorSubject. Then in your SelectMany you can subscribe to the PreviousValue, do whatever comparisons and operations you want, and then push a value back into 'PreviousValue' with a Multicast.

**Reply by jbscherrer** [reply]
Thank you! This all makes sense. I was actually able to rewrite things so that I didn't need to use subjects. But in the future this may be helpful.
